### ISSUE TITLE
variable username/hostname

### ISSUE THREAD

**Issue by ray-holmium** [ne9z/dotfiles-flake#8]
In my journey following this guide I discovered a relatively simple way to allow the user to choose their desired hostname and username during the installation process. It's preferable to prompt the user for these details and use 'sed' to enter the details automatically to the new configuration, rather than having the user change this manually after the system is installed.

I deviated from the guide at section 3.3 (system configuration.customize configuration to hardware), By removing slashes from the 'sed' commands, I replaced every occurrence of "exampleHost" or "exampleUser" from the template:

```
#set hostname
read -rp "Please enter a hostname for this system: " HOSTNAME

#set username
read -rp "Please enter a username: " USERNAME

#apply chosen hostname
cp -r /mnt/etc/nixos/hosts/exampleHost /mnt/etc/nixos/hosts/"${HOSTNAME}"
sed -i "s|"exampleHost"|"${HOSTNAME}"|g" /mnt/etc/nixos/hosts/"${HOSTNAME}"/default.nix
sed -i "s|"exampleHost"|"${HOSTNAME}"|g" /mnt/etc/nixos/flake.nix

#apply chosen username
cp /mnt/etc/nixos/system-modules/users/exampleUser.nix /mnt/etc/nixos/system-modules/users/"${USERNAME}".nix
sed -i "s|"exampleUser"|"${USERNAME}"|g" /mnt/etc/nixos/flake.nix
sed -i "s|"exampleUser"|"${USERNAME}"|g" /mnt/etc/nixos/hosts/"${HOSTNAME}"/default.nix
sed -i "s|"exampleUser"|"${USERNAME}"|g" /mnt/etc/nixos/system-modules/users/default.nix
sed -i "s|"exampleUser"|"${USERNAME}"|g" /mnt/etc/nixos/system-modules/users/"${USERNAME}".nix
sed -i "s|"exampleUser@exampleHost"|"${USERNAME}@${HOSTNAME}"|g" /mnt/etc/nixos/flake.nix
```

**Comment by ne9z** [comment]
Simply put, the less customization, the better.  The goal is to get the
system installed in the shortest time possible, with the least steps
possible.  In the beginning of the guide, I wrote:

> Unless stated otherwise, it is not recommended to customize system
> configuration before reboot.

This is a warranty waiver.  That is, by following the guide, I can get
your computer to boot, from ZFS, with the distro of choice.

Not anything more, not anything less.  Any errors caused by your own
customization is not my responsibility.

Every additional paragraph of text, represents a step which people might
botch the entire endeavor.  Do not overestimate people's ability to
follow instructions, and do not underestimate people's ability to get
creative in how to misintepret your instructions.

For example, some people might want to set the user name to 🐻‍❄️, and I
will be unable to foresee the consequences of doing so.  Will it work?
I don't know.  Do I want to answer such issue reports?  Certainly not.

Before introducing new customization, I ask: will the computer boot
without it?  If it does, then it is not necessary.  Simple as that.
Indeed, "exampleHost" is there because the flake requires it.  If not
for the flake, I would be totally fine with the default. Which is
"nixos".

In short, the proposed changes do not affect the computers ability to
boot.  Therefore I do not think they should be added.  Of course, you
can write your own guide with them included.  The guides are all
licensed under CC-BY.

**Comment by ray-holmium** [comment]
This is rational. By leaving the selection of username/hostname as a post-install step, the guide is absolved of any potential bugbears in user entry (possibly even the literal one you proposed--I didn't think of that).

**Comment by ne9z** [comment]
In general, I think it is not worth your valuable time tweaking NixOS.

For something truly worthwhile, you can learn some logic.  At least I am
really interested in it, but don't have time to read the books.

To Truth Through Proof, by P. B. Andrews.

    https://gtps.math.cmu.edu/tttp.html

Intro to Haskell, by R. Bird.

    https://www.cs.ox.ac.uk/publications/books/functional/

Finally, if you are done with the above, get involved with Isabelle.

    https://isabelle.in.tum.de/

You can read this proof if you are familiar with Linear Algebra, for
example:

    https://isabelle.in.tum.de/dist/library/HOL/HOL-Analysis/Linear_Algebra.html

To quote G. H. Hardy:

> A mathematician, like a painter or poet, is a maker of patterns. If
> his patterns are more permanent than theirs, it is because they are
> made with ideas.

> Archimedes will be remembered when Aeschylus is forgotten, because
> languages die and mathematical ideas do not. "Immortality" may be a
> silly word, but probably a mathematician has the best chance of
> whatever it may mean.

**Comment by ray-holmium** [comment]
Hey, thanks for sharing!!!! This is greatly appreciated. You've sure given me a load of homework--I'll be lucky to get through half. My mathematics skills are very poor. Before I start on these resources, I better first finishing reading this book my friend lent me,  "A Mind For Numbers: How to Excel at Math and Science (Even If You Flunked Algebra)"  by Barbara Oakley. haha!

But your point is well-taken.
